File: app/services/notification_service.py
"""
Notification service (+ PART5 multimedia support).

Sends notifications to users via Telegram.
"""

from typing import Any

from aiogram import Bot
from loguru import logger
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.failed_notification import FailedNotification
from app.repositories.admin_repository import AdminRepository
from app.repositories.failed_notification_repository import (
    FailedNotificationRepository,
)
from app.repositories.support_ticket_repository import SupportTicketRepository


class NotificationService:
    """
    Notification service.

    Handles Telegram notifications with multimedia support (PART5).
    """

    def __init__(self, session: AsyncSession) -> None:
        """Initialize notification service."""
        self.session = session
        self.failed_repo = FailedNotificationRepository(session)
        self.admin_repo = AdminRepository(session)
        self.ticket_repo = SupportTicketRepository(session)

    async def send_notification(
        self,
        bot: Bot,
        user_telegram_id: int,
        message: str,
        critical: bool = False,
        redis_client: Any | None = None,
    ) -> bool:
        """
        Send text notification.

        R11-3: If Redis is unavailable, writes to PostgreSQL fallback queue.

        Args:
            bot: Bot instance
            user_telegram_id: Telegram user ID
            message: Message text
            critical: Mark as critical
            redis_client: Optional Redis client for checking availability

        Returns:
            True if sent successfully or queued to fallback
        """
        # R11-3: Check if Redis is available
        redis_available = False
        if redis_client is not None:
            try:
                await redis_client.ping()
                redis_available = True
            except Exception:
                redis_available = False
                logger.warning(
                    "R11-3: Redis unavailable, will use PostgreSQL fallback"
                )

        # R11-3: If Redis is unavailable, write to PostgreSQL fallback
        if not redis_available:
            try:
                from app.models.notification_queue_fallback import (
                    NotificationQueueFallback,
                )
                from app.repositories.user_repository import UserRepository

                user_repo = UserRepository(self.session)
                user = await user_repo.get_by_telegram_id(user_telegram_id)

                if user:
                    # Create fallback queue entry
                    fallback_entry = NotificationQueueFallback(
                        user_id=user.id,
                        notification_type="text",
                        payload={
                            "message": message,
                            "critical": critical,
                        },
                        priority=100 if critical else 0,
                    )
                    self.session.add(fallback_entry)
                    await self.session.flush()

                    logger.info(
                        f"R11-3: Notification queued to PostgreSQL fallback "
                        f"for user {user_telegram_id} (user_id={user.id})"
                    )
                    return True
                else:
                    logger.warning(
                        f"R11-3: Cannot queue notification for unknown user "
                        f"{user_telegram_id}"
                    )
            except Exception as fallback_error:
                logger.error(
                    f"R11-3: Failed to write to PostgreSQL fallback: {fallback_error}",
                    exc_info=True,
                )

        try:
            await bot.send_message(
                chat_id=user_telegram_id, text=message
            )
            
            # R8-2: If message sent successfully, check if user was previously blocked
            # and reset the flag (user unblocked the bot)
            try:
                from app.repositories.user_repository import UserRepository
                user_repo = UserRepository(self.session)
                user = await user_repo.get_by_telegram_id(user_telegram_id)
                if user and hasattr(user, 'bot_blocked') and user.bot_blocked:
                    # User unblocked the bot - reset flag
                    await user_repo.update(user.id, bot_blocked=False)
                    logger.info(
                        f"User {user_telegram_id} unblocked the bot, flag reset"
                    )
            except Exception as reset_error:
                # Don't fail notification if flag reset fails
                logger.warning(f"Failed to reset bot_blocked flag: {reset_error}")
            
            return True
        except Exception as e:
            # R8-2: Improved 403 error handling with specific TelegramAPIError check
            from aiogram.exceptions import TelegramAPIError
            from datetime import UTC, datetime
            
            # Check for specific "bot was blocked by the user" error
            is_bot_blocked = False
            if isinstance(e, TelegramAPIError):
                # Check error code and message
                if e.error_code == 403:
                    error_message = str(e).lower()
                    if "bot was blocked by the user" in error_message or "blocked" in error_message:
                        is_bot_blocked = True
            else:
                # Fallback for non-TelegramAPIError exceptions
                error_str = str(e).lower()
                if "403" in error_str or "forbidden" in error_str:
                    if "blocked" in error_str or "bot was blocked" in error_str:
                        is_bot_blocked = True
            
            if is_bot_blocked:
                logger.warning(
                    f"Bot blocked by user {user_telegram_id}",
                    extra={"user_id": user_telegram_id},
                )
                
                # Mark user as having blocked bot
                try:
                    from app.repositories.user_repository import UserRepository
                    
                    user_repo = UserRepository(self.session)
                    user = await user_repo.find_by_telegram_id(user_telegram_id)
                    if user and not user.bot_blocked:
                        await user_repo.update(
                            user.id,
                            bot_blocked=True,
                            bot_blocked_at=datetime.now(UTC),
                        )
                        await self.session.commit()
                        logger.info(
                            f"Marked user {user_telegram_id} as bot_blocked"
                        )
                except Exception as update_error:
                    logger.error(
                        f"Failed to mark user as bot_blocked: {update_error}"
                    )
                
                # Don't save to failed notifications for blocked users
                # (they won't receive it anyway)
                return False
            
            logger.error(
                f"Failed to send notification: {e}",
                extra={"user_id": user_telegram_id},
            )

            # Save to failed notifications (PART5) for other errors
            await self._save_failed_notification(
                user_telegram_id,
                "text_message",
                message,
                str(e),
                critical,
            )
            return False

    async def send_photo(
        self,
        bot: Bot,
        user_telegram_id: int,
        file_id: str,
        caption: str | None = None,
    ) -> bool:
        """
        Send photo notification (PART5 multimedia).

        Args:
            bot: Bot instance
            user_telegram_id: Telegram user ID
            file_id: Telegram file ID
            caption: Photo caption

        Returns:
            True if sent successfully
        """
        try:
            await bot.send_photo(
                chat_id=user_telegram_id,
                photo=file_id,
                caption=caption,
            )
            return True
        except Exception as e:
            await self._save_failed_notification(
                user_telegram_id,
                "photo",
                caption or "",
                str(e),
                metadata={"file_id": file_id},
            )
            return False

    async def _save_failed_notification(
        self,
        user_telegram_id: int,
        notification_type: str,
        message: str,
        error: str,
        critical: bool = False,
        metadata: dict[str, Any] | None = None,
    ) -> FailedNotification:
        """Save failed notification for retry (PART5)."""
        return await self.failed_repo.create(
            user_telegram_id=user_telegram_id,
            notification_type=notification_type,
            message=message,
            last_error=error,
            critical=critical,
            notification_metadata=metadata,
        )

    async def notify_admins_new_ticket(
        self, bot: Bot, ticket_id: int
    ) -> None:
        """
        Notify all admins about new support ticket.

        Args:
            bot: Bot instance
            ticket_id: Support ticket ID
        """
        # Get ticket details
        ticket = await self.ticket_repo.get_by_id(ticket_id)
        if not ticket:
            logger.error(
                "Ticket not found for admin notification",
                extra={"ticket_id": ticket_id},
            )
            return

        # Get all admins
        all_admins = await self.admin_repo.find_by()

        if not all_admins:
            logger.warning("No admins found to notify about new ticket")
            return

        # Build notification message
        message = f"""
🆕 **Новое обращение в поддержку**

📋 Тикет #{ticket_id}
👤 Пользователь ID: {ticket.user_id}
📂 Категория: {ticket.category}
🕐 Время создания: {ticket.created_at.strftime('%Y-%m-%d %H:%M:%S')}

Перейдите в админ-панель для обработки.
        """.strip()

        # Send to all admins
        for admin in all_admins:
            try:
                await bot.send_message(
                    chat_id=admin.telegram_id,
                    text=message,
                    parse_mode="Markdown",
                )
                logger.info(
                    "Admin notified about new ticket",
                    extra={
                        "admin_id": admin.id,
                        "ticket_id": ticket_id,
                    },
                )
            except Exception as e:
                logger.error(
                    f"Failed to notify admin about ticket: {e}",
                    extra={
                        "admin_id": admin.id,
                        "ticket_id": ticket_id,
                    },
                )
                await self._save_failed_notification(
                    admin.telegram_id,
                    "admin_notification",
                    message,
                    str(e),
                    critical=True,
                )

    async def notify_admins(
        self,
        bot: Bot,
        message: str,
        critical: bool = False,
    ) -> int:
        """
        Notify all admins with a message.

        Args:
            bot: Bot instance
            message: Message text to send
            critical: Mark as critical notification

        Returns:
            Number of admins successfully notified
        """
        # Get all admins
        all_admins = await self.admin_repo.find_by()

        if not all_admins:
            logger.warning("No admins found to notify")
            return 0

        success_count = 0

        # Send to all admins
        for admin in all_admins:
            try:
                await bot.send_message(
                    chat_id=admin.telegram_id,
                    text=message,
                    parse_mode="Markdown",
                )
                success_count += 1
                logger.info(
                    "Admin notified",
                    extra={
                        "admin_id": admin.id,
                        "telegram_id": admin.telegram_id,
                        "critical": critical,
                    },
                )
            except Exception as e:
                logger.error(
                    f"Failed to notify admin: {e}",
                    extra={
                        "admin_id": admin.id,
                        "telegram_id": admin.telegram_id,
                    },
                )
                await self._save_failed_notification(
                    admin.telegram_id,
                    "admin_notification",
                    message,
                    str(e),
                    critical=critical,
                )

        return success_count

    async def notify_withdrawal_processed(
        self, telegram_id: int, amount: float, tx_hash: str
    ) -> bool:
        """
        Notify user about withdrawal being processed.

        Args:
            telegram_id: User telegram ID
            amount: Withdrawal amount
            tx_hash: Transaction hash

        Returns:
            True if notification sent successfully
        """
        from bot.main import bot_instance
        from aiogram import Bot
        from aiogram.client.default import DefaultBotProperties
        from aiogram.enums import ParseMode
        from app.config.settings import settings

        bot = bot_instance
        should_close = False

        if not bot:
            try:
                bot = Bot(
                    token=settings.telegram_bot_token,
                    default=DefaultBotProperties(parse_mode=ParseMode.MARKDOWN),
                )
                should_close = True
            except Exception as e:
                logger.error(f"Failed to create fallback bot instance: {e}")
                return False

        message = (
            f"✅ **Выплата отправлена!**\n\n"
            f"💰 Сумма: {amount:.2f} USDT\n"
            f"🔗 TX: [Посмотреть транзакцию](https://bscscan.com/tx/{tx_hash})\n\n"
            f"🤝 Спасибо за ваше доверие к SigmaTrade!"
        )

        try:
            await bot.send_message(
                chat_id=telegram_id,
                text=message,
                parse_mode="Markdown",
                disable_web_page_preview=True
            )
            return True
        except Exception as e:
            logger.error(
                f"Failed to notify user about withdrawal: {e}",
                extra={"telegram_id": telegram_id},
            )
            return False
        finally:
            if should_close and bot:
                await bot.session.close()

    async def notify_withdrawal_rejected(
        self, telegram_id: int, amount: float
    ) -> bool:
        """
        Notify user about withdrawal being rejected.

        Args:
            telegram_id: User telegram ID
            amount: Withdrawal amount

        Returns:
            True if notification sent successfully
        """
        from bot.main import bot_instance
        from aiogram import Bot
        from aiogram.client.default import DefaultBotProperties
        from aiogram.enums import ParseMode
        from app.config.settings import settings

        bot = bot_instance
        should_close = False

        if not bot:
            try:
                bot = Bot(
                    token=settings.telegram_bot_token,
                    default=DefaultBotProperties(parse_mode=ParseMode.MARKDOWN),
                )
                should_close = True
            except Exception as e:
                logger.error(f"Failed to create fallback bot instance: {e}")
                return False

        message = (
            f"❌ **Ваша заявка на вывод средств отклонена**\n\n"
            f"💰 Сумма: {amount:.2f} USDT\n\n"
            f"Для получения дополнительной информации обратитесь в поддержку."
        )

        try:
            await bot.send_message(
                chat_id=telegram_id,
                text=message,
                parse_mode="Markdown",
            )
            return True
        except Exception as e:
            logger.error(
                f"Failed to notify user about withdrawal rejection: {e}",
                extra={"telegram_id": telegram_id},
            )
            return False
        finally:
            if should_close and bot:
                await bot.session.close()

    async def notify_roi_accrual(
        self,
        telegram_id: int,
        amount: float,
        deposit_level: int,
        roi_progress_percent: float,
    ) -> bool:
        """
        Notify user about ROI accrual.

        Args:
            telegram_id: User telegram ID
            amount: ROI amount accrued
            deposit_level: Deposit level
            roi_progress_percent: Current ROI progress (0-500%)

        Returns:
            True if notification sent successfully
        """
        from bot.main import bot_instance
        from aiogram import Bot
        from aiogram.client.default import DefaultBotProperties
        from aiogram.enums import ParseMode
        from app.config.settings import settings

        bot = bot_instance
        should_close = False

        if not bot:
            try:
                bot = Bot(
                    token=settings.telegram_bot_token,
                    default=DefaultBotProperties(parse_mode=ParseMode.MARKDOWN),
                )
                should_close = True
            except Exception as e:
                logger.error(f"Failed to create fallback bot instance: {e}")
                return False

        # Progress bar
        filled = int(roi_progress_percent / 50)  # 10 blocks for 500%
        empty = 10 - filled
        progress_bar = "█" * filled + "░" * empty

        message = (
            f"💰 *Начислен ROI*\n\n"
            f"📊 Уровень {deposit_level}: *+{amount:.2f} USDT*\n"
            f"📈 Прогресс: {progress_bar} {roi_progress_percent:.1f}%\n\n"
            f"_Баланс обновлён автоматически._"
        )

        try:
            await bot.send_message(
                chat_id=telegram_id,
                text=message,
                parse_mode="Markdown",
            )
            return True
        except Exception as e:
            logger.error(
                f"Failed to notify user about ROI accrual: {e}",
                extra={"telegram_id": telegram_id},
            )
            return False
        finally:
            if should_close and bot:
                await bot.session.close()

File: .git-rewrite/t/src/services/notification.service.ts
/**
 * Notification Service
 * Handles sending notifications to users via Telegram
 * - Deposit confirmations
 * - Referral earnings
 * - Level activations
 * - System alerts
 */

import { Telegraf } from 'telegraf';
import { logger } from '../utils/logger.util';

export class NotificationService {
  private static instance: NotificationService;
  private bot?: Telegraf;

  private constructor() {}

  public static getInstance(): NotificationService {
    if (!NotificationService.instance) {
      NotificationService.instance = new NotificationService();
    }
    return NotificationService.instance;
  }

  /**
   * Set bot instance (called from bot initialization)
   */
  public setBot(bot: Telegraf): void {
    this.bot = bot;
  }

  /**
   * Send notification to user
   */
  private async sendNotification(
    telegramId: number,
    message: string,
    options?: { parse_mode?: 'Markdown' | 'HTML' }
  ): Promise<boolean> {
    if (!this.bot) {
      logger.error('Bot not initialized in NotificationService');
      return false;
    }

    try {
      await this.bot.telegram.sendMessage(telegramId, message, options);
      return true;
    } catch (error) {
      logger.error('Error sending notification', {
        telegramId,
        error: error instanceof Error ? error.message : String(error),
      });
      return false;
    }
  }

  /**
   * Notify user about deposit confirmation
   */
  public async notifyDepositConfirmed(
    telegramId: number,
    amount: number,
    level: number,
    txHash: string
  ): Promise<void> {
    const message = `
✅ **Депозит подтвержден!**

💰 Сумма: ${amount} USDT
📊 Уровень: ${level}
🔗 Транзакция: \`${txHash}\`

Ваш депозит успешно подтвержден в блокчейне BSC!
Уровень ${level} активирован. 🎉

[Посмотреть в BSCScan](https://bscscan.com/tx/${txHash})
    `.trim();

    await this.sendNotification(telegramId, message, { parse_mode: 'Markdown' });

    logger.info('Deposit confirmation notification sent', {
      telegramId,
      amount,
      level,
    });
  }

  /**
   * Notify user about referral earning
   */
  public async notifyReferralEarning(
    telegramId: number,
    amount: number,
    level: number,
    referredUsername?: string
  ): Promise<void> {
    const levelNames = {
      1: 'прямого реферала',
      2: 'реферала 2 уровня',
      3: 'реферала 3 уровня',
    };

    const referredInfo = referredUsername
      ? `от ${referredUsername}`
      : `уровня ${level}`;

    const message = `
💵 **Получено реферальное вознаграждение!**

💰 Сумма: ${amount} USDT
👥 От: ${levelNames[level as keyof typeof levelNames] || `уровня ${level}`}
${referredUsername ? `👤 Реферал: @${referredUsername}` : ''}

Вознаграждение будет выплачено на ваш кошелек автоматически.
    `.trim();

    await this.sendNotification(telegramId, message, { parse_mode: 'Markdown' });

    logger.info('Referral earning notification sent', {
      telegramId,
      amount,
      level,
    });
  }

  /**
   * Notify user about payment sent
   */
  public async notifyPaymentSent(
    telegramId: number,
    amount: number,
    txHash: string
  ): Promise<void> {
    const message = `
💸 **Выплата отправлена!**

💰 Сумма: ${amount} USDT
🔗 Транзакция: \`${txHash}\`

Реферальные вознаграждения отправлены на ваш кошелек!

[Посмотреть в BSCScan](https://bscscan.com/tx/${txHash})
    `.trim();

    await this.sendNotification(telegramId, message, { parse_mode: 'Markdown' });

    logger.info('Payment sent notification sent', {
      telegramId,
      amount,
    });
  }

  /**
   * Notify user about new referral
   */
  public async notifyNewReferral(
    telegramId: number,
    referralUsername?: string
  ): Promise<void> {
    const referralInfo = referralUsername
      ? `**@${referralUsername}**`
      : 'Новый пользователь';

    const message = `
🎉 **У вас новый реферал!**

👤 ${referralInfo} зарегистрировался по вашей ссылке!

Вы будете получать вознаграждения от его депозитов:
• 3% от прямых депозитов
• 2% от депозитов его рефералов
• 5% от депозитов рефералов 3 уровня

Продолжайте приглашать друзей! 💰
    `.trim();

    await this.sendNotification(telegramId, message, { parse_mode: 'Markdown' });

    logger.info('New referral notification sent', {
      telegramId,
      referralUsername,
    });
  }

  /**
   * Notify user about level activation
   */
  public async notifyLevelActivated(
    telegramId: number,
    level: number
  ): Promise<void> {
    const message = `
🎊 **Уровень ${level} активирован!**

Поздравляем! Вы успешно активировали уровень ${level}.

${level < 5 ? `Следующий уровень: ${level + 1}\nТребуется рефералов: ${level}` : 'Вы достигли максимального уровня! 🏆'}
    `.trim();

    await this.sendNotification(telegramId, message, { parse_mode: 'Markdown' });

    logger.info('Level activation notification sent', {
      telegramId,
      level,
    });
  }

  /**
   * Notify admin about system event
   */
  public async notifyAdmin(
    adminTelegramId: number,
    title: string,
    message: string
  ): Promise<void> {
    const fullMessage = `
🔔 **${title}**

${message}
    `.trim();

    await this.sendNotification(adminTelegramId, fullMessage, {
      parse_mode: 'Markdown',
    });

    logger.info('Admin notification sent', {
      adminTelegramId,
      title,
    });
  }

  /**
   * Notify user about deposit pending
   */
  public async notifyDepositPending(
    telegramId: number,
    amount: number,
    level: number,
    txHash: string
  ): Promise<void> {
    const message = `
⏳ **Депозит обнаружен!**

💰 Сумма: ${amount} USDT
📊 Уровень: ${level}
🔗 Транзакция: \`${txHash}\`

Ожидаем подтверждение в блокчейне (12 блоков).
Обычно это занимает 1-2 минуты.

[Отследить транзакцию](https://bscscan.com/tx/${txHash})
    `.trim();

    await this.sendNotification(telegramId, message, { parse_mode: 'Markdown' });

    logger.info('Deposit pending notification sent', {
      telegramId,
      amount,
      level,
    });
  }

  /**
   * Notify user about deposit timeout
   */
  public async notifyDepositTimeout(
    telegramId: number,
    amount: number,
    level: number
  ): Promise<void> {
    const message = `
⏱️ **Время ожидания депозита истекло**

💰 Сумма: ${amount} USDT
📊 Уровень: ${level}

К сожалению, мы не обнаружили депозит в течение 24 часов.

Возможные причины:
• Транзакция не была отправлена
• Неправильный адрес или сеть
• Недостаточная сумма для покрытия комиссии

Если вы отправили средства, свяжитесь с поддержкой.
    `.trim();

    await this.sendNotification(telegramId, message, { parse_mode: 'Markdown' });

    logger.info('Deposit timeout notification sent', {
      telegramId,
      amount,
      level,
    });
  }

  /**
   * Notify user about withdrawal request received
   */
  public async notifyWithdrawalReceived(
    telegramId: number,
    amount: number
  ): Promise<void> {
    const message = `
📤 **Заявка на вывод получена**

💰 Сумма: ${amount} USDT

Ваша заявка принята в обработку.
Обычно вывод занимает 15-30 минут.

Мы уведомим вас, когда средства будут отправлены.
    `.trim();

    await this.sendNotification(telegramId, message, { parse_mode: 'Markdown' });

    logger.info('Withdrawal received notification sent', {
      telegramId,
      amount,
    });
  }

  /**
   * Notify user about withdrawal processed
   */
  public async notifyWithdrawalProcessed(
    telegramId: number,
    amount: number,
    txHash: string
  ): Promise<void> {
    const message = `
✅ **Вывод выполнен!**

💰 Сумма: ${amount} USDT
🔗 Транзакция: \`${txHash}\`

Средства отправлены на ваш кошелек!

[Посмотреть в BSCScan](https://bscscan.com/tx/${txHash})
    `.trim();

    await this.sendNotification(telegramId, message, { parse_mode: 'Markdown' });

    logger.info('Withdrawal processed notification sent', {
      telegramId,
      amount,
    });
  }
}

// Export singleton instance
export const notificationService = NotificationService.getInstance();
